[RDY] LSP: only send buf requests to servers that support the request

fsouza opened the pull request:
Refactors how required capabilities are detected and validated, and make sure requests are only sent to clients that support it (and only fail if no clients support the provided method).

The validation happens at the buf_request level, because we assume that if someone is sending the request directly through the client, they know what they're doing. Also, let unknown methods go through.

Let me know if you'd like me to squash the changes, I basically brought the relevant commits from the other PR.

This is extracted from #12518 and closes #12755.

tjdevries commented on runtime/lua/vim/lsp.lua:
```diff
@@ -25,13 +25,40 @@ local lsp = {
 end
 
+local function unsupported_method(method)
+  local msg = string.format("server doesn't support '%s'", method)
```
Is there any way to get what server this is (in the case of multiple servers for one buffer?)

fsouza replied:
With this change, this is only being called if none of the servers support the method. Should we list all servers?

I've updated the message to be more explicit, let me know if that's too verbose.

tjdevries commented on runtime/lua/vim/lsp/buf.lua:
```diff
@@ -141,7 +142,7 @@ end
 --- provided, the user will be prompted for a new name using |input()|.
 function M.rename(new_name)
-  -- TODO(ashkan) use prepareRename
+ -- TODO(ashkan) use prepareRename
```
whitespace change?

fsouza replied:
my bad. fixed

tjdevries commented on runtime/lua/vim/lsp/buf.lua:
```diff
@@ -15,12 +15,13 @@ local function npcall(fn, ...)
+    unsupported_callback = {unsupported_callback, 'f', true};
   }
-  return vim.lsp.buf_request(0, method, params, callback)
+  return vim.lsp.buf_request(0, method, params, callback, unsupported_callback)
```
I think you also need to change `vim.lsp.buf_request` to accept `unsupported_callback`. As far as I can tell right now it is not used at all (just passes an extra arg to that function).

Also needs doc updates to add it.

fsouza replied:
I think this was left-over from a previous approach. Removed.

tjdevries commented:
Why does this need the new `unsupported_callback` param? I think your request passes a callback that gets the `err` part of the message.

The `callback` taks `err` as the first argument. So it should already be able to do whatever you want w/ `unsupported_callback` that you'd do anyway.

Does that make sense?

tjdevries commented:
(Also, could you fix the conflicts from master?)

justinmk commented on runtime/lua/vim/lsp.lua:
```diff
@@ -25,13 +25,40 @@ local lsp = {
 }
 
+-- maps request name to the required resolved_capability in the client.
+local req_to_capability = {
```
(unrelated to this PR) why do we have different key names that require this mapping? why don't we use the protocol-defined names?

fsouza replied:
not sure if I follow the comment, sorry about that. Do you mean introducing some shared variables for the method names or something like that? Or if the protocol itself already declares that dependency-mapping? Or something else? x)

tjdevries replied:
These are the protocol defined names, I think it's fine.

fsouza commented:
@tjdevries I've finally come across to address comments on this PR. I think it's ready for review again. I promise a faster turnaround if yall have more feedback :)

> Why does this need the new `unsupported_callback` param? I think your request passes a callback that gets the `err` part of the message.

I think this was left-over from the other branch or something else I tried, sorry about that.

fsouza commented:
(also let me know if you want me to squash the commits)

tjdevries commented on runtime/lua/vim/lsp.lua:
```diff
@@ -50,6 +71,16 @@ local function resolve_bufnr(bufnr)
+--- callback called by the client when trying to call a method that's not
+--- supported in any of the servers registered for the current buffer.
+--@param method (string) name of the method
+local function unsupported_method(method)
```
let's make this not local.

Then people can just override it.

`vim.lsp._unsupported_method` or something like that maybe? That way people can easily override the functionality (or silence the errors if they don't care)

fsouza replied:
Wouldn't that make it part of the public API and difficult to change in the future? If we do end-up making it public, should it be part of the callbacks module? Or should we keep that module for callbacks to methods that exist in the LSP spec?

tjdevries commented:
Last thing would be, do you think any tests we should add for this?

fsouza commented:
> Last thing would be, do you think any tests we should add for this?

I think I can copy one of the existing tests and emulate the server not supporting the used method and check that unsupported_method is called or something like that.

I can also review the current tests to ensure they exercise the client capabilities check. What do you think?

fsouza commented:
@tjdevries I added some tests and fixed a bug that would happen when the callback is nil (I have been running this branch, but I wasn't able to trigger the bug because I detect capabilities before setting up the maps, I was only able to catch the bug with the test!)

I haven't squashed my commits yet to make it easier for you to review, but let me know if you want me to do it or if you have any other feedback! :)

fsouza commented:
@tjdevries it looks like CI is green 😁

fsouza commented:
@tjdevries I squashed all commits prior to your  review in two, and can squash again once it's approved.

fsouza committed:
lsp: only send buf requests to servers that support the request

Refactors how required capabilities are detected and validated, and make
sure requests are only sent to clients that support it (and only fail if
no clients support the provided method).

The validation happens at the buf_request level, because we assume that
if someone is sending the request directly through the client, they know
what they're doing. Also, let unknown methods go through.

This is extracted from #12518 and closes #12755.

tjdevries commented on runtime/lua/vim/lsp.lua:
```diff
@@ -50,6 +71,16 @@ local function resolve_bufnr(bufnr)
+--@param method (string) name of the method
+function lsp._unsupported_method(method)
+  local msg = string.format("method %s is not supported by any of the servers registered for the current buffer", method)
+  local _ = log.warn() and log.warn(msg)
```
you can just do `log.warn(msg)` -- you only need to do the check if it is expensive to calculate `msg`.

tjdevries commented on runtime/lua/vim/lsp.lua:
```diff
@@ -25,6 +25,27 @@ local lsp = {
 }
 
+-- maps request name to the required resolved_capability in the client.
+local req_to_capability = {
```
I think maybe we should expose this as `vim.lsp._request_name_to_capability` -- so people can add different methods to check if they need to later (and not just to default to true).

fsouza replied:
Perhaps we can ship as-is and see if there's demand for exposing it? I'm assuming that removing something from the public API is harder than adding it 🙈 (or does the `_` prefix gives us some freedom to remove it later if we figure it's not used).

tjdevries replied:
`_` prefix means "use at your own risk".

tjdevries replied:
(so we can change whenever and however we want, particularly before we release)

At this point, if we have a good reason, we can break "compatibility" with earlier 0.5 items because we haven't released 0.5 yet anyway (even for non "private" items). That is what happens when you follow master sometimes (rarely, but it does happen)

fsouza replied:
Sounds good, will adjust then!

fsouza replied:
Done! 🎉

tjdevries commented:
OK, if you agree with the things I mentioned and you fix those up, I will merge.

Thanks for your patience.
